firedrake/preconditioners/pmg.py

```diff
@@ -274,55 +295,236 @@ def __init__(self, Vf, Vc, Vf_bcs, Vc_bcs):
 
         self.uc = firedrake.Function(Vc)
         self.uf = firedrake.Function(Vf)
-        self.prolong_kernel = self.prolongation_transfer_kernel_action(Vf, self.uc)
 
+        self.mesh = Vf.mesh()
+        self.weight = self.multiplicity(Vf)
+        with self.weight.dat.vec as w:
+            w.reciprocal()
+
+        tf, _, _ = isTensorProductSpace(Vf)
+        tc, _, _ = isTensorProductSpace(Vc)
+        if tf and tc:
+            self.make_blas_kernels(Vf, Vc)
+        else:
+            self.make_kernels(Vf, Vc)
+        return
+
+    def make_kernels(self, Vf, Vc):
+        # The way we transpose the prolongation kernel is suboptimal, but works for general elements.
+        # A local matrix has to be generated each time the kernel is executed.
+        # We are waiting for structure-preserving tfsc kernels.
+        # As an alternative we provide blas kernels for tensor product elements.
+        self.prolong_kernel = self.prolongation_transfer_kernel_action(Vf, self.uc)
         matrix_kernel = self.prolongation_transfer_kernel_action(Vf, firedrake.TestFunction(Vc))
         element_kernel = loopy.generate_code_v2(matrix_kernel.code).device_code()
         element_kernel = element_kernel.replace("void expression_kernel", "static void expression_kernel")
         dimc = Vc.finat_element.space_dimension() * Vc.value_size
         dimf = Vf.finat_element.space_dimension() * Vf.value_size
-        self.restrict_code = f"""
-{element_kernel}
+        restrict_code = f"""
+        {element_kernel}
+
+        void restriction(double *restrict Rc, const double *restrict Rf, const double *restrict w)
+        {{
+            double Afc[{dimf}*{dimc}] = {{0}};
+            expression_kernel(Afc);
+            for (int32_t i = 0; i < {dimf}; i++)
+               for (int32_t j = 0; j < {dimc}; j++)
+                   Rc[j] += Afc[i*{dimc} + j] * Rf[i] * w[i];
+        }}
+        """
+        self.restrict_kernel = op2.Kernel(restrict_code, "restriction")
+        return
 
+    @staticmethod
+    def prolongation_transfer_kernel_action(Vf, expr):
+        from tsfc import compile_expression_dual_evaluation
+        from tsfc.fiatinterface import create_element
+        coords = Vf.ufl_domain().coordinates
+        to_element = create_element(Vf.ufl_element(), vector_is_mixed=False)
+        ast, oriented, needs_cell_sizes, coefficients, _ = compile_expression_dual_evaluation(expr, to_element, coords, coffee=False)
+        return op2.Kernel(ast, ast.name)
+
+    def make_blas_kernels(self, Vf, Vc):
```

Comment:
Could you add an outline comment, explaining the main difference between these kernels and the previous ones?

Reply:
And ideally a reference to somewhere (which might e.g. be a transfer thesis) where the maths behind it is worked out.